modules/network/src/main/java/org/apache/ignite/internal/network/recovery/RecoveryServerHandshakeManager.java

```diff
@@ -148,24 +170,51 @@ public void onMessage(NetworkMessage message) {
         if (message instanceof HandshakeStartResponseMessage) {
             HandshakeStartResponseMessage msg = (HandshakeStartResponseMessage) message;
 
-            if (staleIdDetector.isIdStale(msg.launchId().toString())) {
+            UUID remoteLaunchId = msg.launchId();
+            String remoteConsistentId = msg.consistentId();
+            long remoteReceivedCount = msg.receivedCount();
+            short remoteChannelId = msg.connectionId();
+
+            if (staleIdDetector.isIdStale(remoteLaunchId.toString())) {
                 handleStaleClientId(msg);
 
                 return;
             }
 
-            this.remoteLaunchId = msg.launchId();
-            this.remoteConsistentId = msg.consistentId();
-            this.receivedCount = msg.receivedCount();
-            this.remoteChannelId = msg.connectionId();
+            this.remoteLaunchId = remoteLaunchId;
+            this.remoteConsistentId = remoteConsistentId;
+            this.receivedCount = remoteReceivedCount;
+            this.remoteChannelId = remoteChannelId;
+
+            RecoveryDescriptor descriptor = recoveryDescriptorProvider.getRecoveryDescriptor(
+                    this.remoteConsistentId,
+                    this.remoteLaunchId,
+                    this.remoteChannelId
+            );
+
+            while (!descriptor.acquire(ctx)) {
+                if (launchId.compareTo(remoteLaunchId) <= 0) {
```

Comment:
Please add a comment or clarify this by extracting a method

Reply:
Done